File: ProjectAndroidStudio/MySQLdatabase/app/src/main/java/com/komputerkit/mysqldatabase/DatabaseHelper.java
package com.komputerkit.mysqldatabase;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

public class DatabaseHelper extends SQLiteOpenHelper {
    // Nama database dan versi
    public static final String DATABASE_NAME = "dbtoko.db";
    public static final int DATABASE_VERSION = 1;

    // Nama tabel dan kolom
    public static final String TABLE_NAME = "tblbarang";
    public static final String COL_ID = "ID_barang";
    public static final String COL_NAMA = "nama_barang";
    public static final String COL_STOK = "stok";
    public static final String COL_HARGA = "harga";

    public DatabaseHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
    }

    // Membuat tabel saat database pertama kali dibuat
    @Override
    public void onCreate(SQLiteDatabase db) {
        String createTable = "CREATE TABLE " + TABLE_NAME + " (" +
                COL_ID + " INTEGER PRIMARY KEY AUTOINCREMENT, " +
                COL_NAMA + " TEXT NOT NULL, " +
                COL_STOK + " INTEGER DEFAULT 0, " +
                COL_HARGA + " REAL DEFAULT 0.0);";
        db.execSQL(createTable);
    }

    // Menangani upgrade database
    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        db.execSQL("DROP TABLE IF EXISTS " + TABLE_NAME);
        onCreate(db);
    }

    // Menjalankan perintah SQL langsung (misal: INSERT manual) dengan penanganan error
    public boolean runSQL(String sql) {
        SQLiteDatabase db = this.getWritableDatabase();
        try {
            db.execSQL(sql);
            return true;
        } catch (android.database.SQLException e) {
            return false;
        }
    }

    // Menambah data barang baru
    public boolean tambahBarang(String nama, int stok, double harga) {
        SQLiteDatabase db = this.getWritableDatabase();
        ContentValues contentValues = new ContentValues();
        contentValues.put(COL_NAMA, nama);
        contentValues.put(COL_STOK, stok);
        contentValues.put(COL_HARGA, harga);
        long result = db.insert(TABLE_NAME, null, contentValues);
        return result != -1;
    }

    /**
     * Mengambil semua data barang dari tabel. Caller WAJIB menutup cursor setelah selesai.
     * Jangan tutup database di sini, biarkan Android yang mengelola lifecycle database.
     */
    public Cursor lihatSemuaBarang() {
        return this.getReadableDatabase().rawQuery("SELECT * FROM " + TABLE_NAME, null);
    }

    // Memperbarui data barang berdasarkan ID
    public boolean updateBarang(String id, String nama, int stok, double harga) {
        SQLiteDatabase db = this.getWritableDatabase();
        ContentValues contentValues = new ContentValues();
        contentValues.put(COL_NAMA, nama);
        contentValues.put(COL_STOK, stok);
        contentValues.put(COL_HARGA, harga);
        int result = db.update(TABLE_NAME, contentValues, COL_ID + " = ?", new String[]{id});
        return result > 0;
    }

    // Menghapus data barang berdasarkan ID
    public Integer hapusBarang(String id) {
        SQLiteDatabase db = this.getWritableDatabase();
        return db.delete(TABLE_NAME, COL_ID + " = ?", new String[]{id});
    }
}

File: ProjectAndroidStudio/MySQLdatabase/app/src/main/java/com/komputerkit/mysqldatabase/DatabaseHelper.kt
package com.komputerkit.mysqldatabase

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.content.ContentValues
import android.database.Cursor

// Data class untuk model siswa
data class Student(val id: Int, val name: String, val studentClass: String)

class DatabaseHelper(context: Context) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {
    companion object {
        const val DATABASE_NAME = "Student.db"
        const val DATABASE_VERSION = 1
        const val TABLE_NAME = "students"
        const val COL_ID = "id"
        const val COL_NAME = "name"
        const val COL_CLASS = "class"
    }

    override fun onCreate(db: SQLiteDatabase) {
        val createTable = "CREATE TABLE $TABLE_NAME (" +
                "$COL_ID INTEGER PRIMARY KEY, " +
                "$COL_NAME TEXT, " +
                "$COL_CLASS TEXT)"
        db.execSQL(createTable)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL("DROP TABLE IF EXISTS $TABLE_NAME")
        onCreate(db)
    }

    // Fungsi untuk membaca semua data siswa
    fun readAllStudents(): List<Student> {
        val students = mutableListOf<Student>()
        val db = this.readableDatabase
        var cursor: Cursor? = null
        try {
            cursor = db.rawQuery("SELECT * FROM $TABLE_NAME", null)
            if (cursor.moveToFirst()) {
                do {
                    val id = cursor.getInt(cursor.getColumnIndexOrThrow(COL_ID))
                    val name = cursor.getString(cursor.getColumnIndexOrThrow(COL_NAME))
                    val studentClass = cursor.getString(cursor.getColumnIndexOrThrow(COL_CLASS))
                    students.add(Student(id, name, studentClass))
                } while (cursor.moveToNext())
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            cursor?.close()
            db.close()
        }
        return students
    }

    // Fungsi untuk menambah data siswa (untuk data statis awal)
    fun insertStudent(student: Student): Boolean {
        val db = this.writableDatabase
        val values = ContentValues().apply {
            put(COL_ID, student.id)
            put(COL_NAME, student.name)
            put(COL_CLASS, student.studentClass)
        }
        val result = db.insert(TABLE_NAME, null, values)
        db.close()
        return result != -1L
    }
}
